### PR TITLE
profiler: Keep track of stack unwinding failures

### PR DESCRIPTION
**Pull request by kakkoyun**
Signed-off-by: Kemal Akkoyun <kakkoyun@gmail.com>

As a pre-work to #346, add metrics to keep track of stack unwinding failures.

cc @v-thakkar

### CODE REVIEW

File: parca-agent.bpf.c [hunk 1]
```diff
@@ -97,9 +100,27 @@ int do_sample(struct bpf_perf_event_data *ctx) {
...
+
+  // get kernel stack id
+  key.kernel_stack_id = 0;
+  int kernel_stack_id = bpf_get_stackid(ctx, &stack_traces, 0);
+  if (kernel_stack_id >= 0)
+    key.kernel_stack_id = kernel_stack_id;
+
+  // TODO(kakkoyun): failed bpf_get_stackid() could indicate stack unwinding
+  // issues; this could be a useful place to hook eh_frame-based stack
+  // unwinding.
```
Comment (v-thakkar) [inline parca-agent.bpf.c:117]: :100:

**Comment by v-thakkar** [comment]
LGTM!

File: parca-agent.bpf.c [hunk 2]
```diff
@@ -97,9 +100,27 @@ int do_sample(struct bpf_perf_event_data *ctx) {
   // create map key
   stack_count_key_t key = {.pid = tgid};
```
Comment (javierhonduco) [inline parca-agent.bpf.c:101]: nit: for consistency, should we initialise the other fields here?

**Review by javierhonduco — COMMENTED** [review]
Looks great!! 

I think you might have mentioned this at some point, but perhaps we could check that the addresses in userspace are all within the stack low<->top addresses? I am unsure if the kernel does this already (checking now!) but if it doesn't it might be worth doing the check ourselves

**Comment by kakkoyun** [comment]
> Looks great!!
> 
> I think you might have mentioned this at some point, but perhaps we could check that the addresses in userspace are all within the stack low<->top addresses? I am unsure if the kernel does this already (checking now!) but if it doesn't it might be worth doing the check ourselves

I don't think kernel does this. I came to that conclusion because we received addresses that we can't map any object file. AFAIK (don't quote me on that), if we receive out-of-bounds addresses, we can say that the Kernel failed to unwind the stack. The only boundary check we have to find the mapping.

https://github.com/parca-dev/parca-agent/blob/e8afc39e6c19f9cb1fa2684a150ae85068d3de6b/pkg/maps/mapping.go#L82-L90

Wed Apr 20 02:19:27 PM CEST 2022

The addresses that we couldn't symbolize and don't have `[*]` prefix is how it manifests itself on icicle graphs.

We should investigate this more, though. Happy to happen an issue (if we haven't gotten one yet)!

**Pull request merged**